File: src/status.ts
/** Query tmux state and format status output. */

import { join } from "@std/path";
import { type ApprovalMode, type Config, sessionName } from "./config.ts";
import type { TaskCounts } from "./task-queue.ts";
import * as tmux from "./tmux.ts";
import { worktreeDir } from "./worktree.ts";

const CURRENT_TASK_DIR = ".jackops/current-task";

function formatTime(epoch: number): string {
  const d = new Date(epoch * 1000);
  return d.toLocaleString([], {
    month: "short",
    day: "numeric",
    hour: "2-digit",
    minute: "2-digit",
  });
}

export type WorkerState = "working" | "waiting" | "stopped" | "gone";

export interface WorkerStatus {
  name: string;
  agent: string;
  state: WorkerState;
  worktree: string;
}

export interface DaemonStatus {
  running: boolean;
}

export type { TaskCounts };

const APPROVAL_DESCRIPTIONS: Record<ApprovalMode, string> = {
  manual: "manual approval required",
  auto: "LLM evaluates permission prompts",
  yolo: "all prompts auto-approved",
};

const SHELLS = new Set(["bash", "zsh", "fish", "sh"]);

export async function getSessionStarted(
  config: Config,
): Promise<number | null> {
  const session = sessionName(config.project);
  if (!(await tmux.hasSession(session))) return null;
  return await tmux.sessionCreated(session);
}

function paneIsRunning(pane: tmux.PaneInfo): boolean {
  const cmd = pane.currentCommand;
  return !pane.paneDead && !!cmd && !SHELLS.has(cmd);
}

async function hasCurrentTask(
  base: string,
  workerName: string,
): Promise<boolean> {
  try {
    await Deno.stat(join(base, CURRENT_TASK_DIR, workerName));
    return true;
  } catch (e) {
    if (e instanceof Deno.errors.NotFound) return false;
    throw e;
  }
}

export async function getStatus(
  config: Config,
  base?: string,
): Promise<{ workers: WorkerStatus[]; daemon: DaemonStatus }> {
  const session = sessionName(config.project);
  if (!(await tmux.hasSession(session))) {
    return { workers: [], daemon: { running: false } };
  }

  const panes = await tmux.listPanes(session);
  const paneByWindow = new Map<string, tmux.PaneInfo>();
  for (const p of panes) {
    paneByWindow.set(p.windowName, p);
  }

  const workers = await Promise.all(config.workers.map(async (w) => {
    const pane = paneByWindow.get(w.name);
    let state: WorkerState = "gone";
    if (pane) {
      if (!paneIsRunning(pane)) {
        state = "stopped";
      } else if (base && await hasCurrentTask(base, w.name)) {
        state = "working";
      } else {
        state = "waiting";
      }
    }
    return {
      name: w.name,
      agent: w.agent,
      state,
      worktree: worktreeDir(config.project, w.name),
    };
  }));

  const dashboard = paneByWindow.get("dashboard-orchestrator");
  const daemon: DaemonStatus = {
    running: !!dashboard && paneIsRunning(dashboard),
  };

  return { workers, daemon };
}

export interface StatusInfo {
  statuses: WorkerStatus[];
  startedEpoch: number | null;
  daemon: DaemonStatus;
  tasks: TaskCounts | null;
  autoApprovalModel?: string | null;
  runtimeApproval?: ApprovalMode | null;
}

export interface JsonStatus {
  session: string;
  daemon: DaemonStatus;
  workers: (WorkerStatus & { branch: string })[];
  tasks: TaskCounts;
}

export function formatJsonStatus(
  config: Config,
  info: StatusInfo,
): JsonStatus {
  const session = sessionName(config.project);
  return {
    session,
    daemon: info.daemon,
    workers: info.statuses.map((s) => ({
      ...s,
      branch: `jackops/${config.project}/${s.name}`,
    })),
    tasks: info.tasks ??
      { pending: 0, current: 0, review: 0, complete: 0, rejected: 0 },
  };
}

export function formatStatus(config: Config, info: StatusInfo): string {
  const { statuses, startedEpoch, daemon, tasks } = info;
  const lines: string[] = [];
  const session = sessionName(config.project);
  const approval = info.runtimeApproval ?? config.orchestrator.approval;

  // Header
  const header = startedEpoch
    ? `JACKOPS -- ${config.project} (started ${formatTime(startedEpoch)})`
    : `JACKOPS -- ${config.project}`;
  lines.push(header);

  if (statuses.length === 0) {
    lines.push("");
    lines.push("  No active session.");
    return lines.join("\n");
  }

  // Session info
  lines.push(`Session:  ${session}`);
  lines.push(`Daemon:   ${daemon.running ? "running" : "stopped"}`);
  lines.push(`Approval: ${approval} (${APPROVAL_DESCRIPTIONS[approval]})`);
  if (approval === "auto" && info.autoApprovalModel) {
    lines.push(`Model:    ${info.autoApprovalModel}`);
  }

  // Workers
  lines.push("");
  lines.push("Workers:");

  const nameWidth = Math.max(...statuses.map((s) => s.name.length), 4);
  const agentWidth = Math.max(...statuses.map((s) => s.agent.length), 5);
  const stateWidth = Math.max(...statuses.map((s) => s.state.length), 6);

  for (const s of statuses) {
    lines.push(
      `  ${s.name.padEnd(nameWidth)}  ${s.agent.padEnd(agentWidth)}  ${
        s.state.padEnd(stateWidth)
      }  ${s.worktree}`,
    );
  }

  // Tasks
  if (tasks) {
    const total = tasks.pending + tasks.current + tasks.review +
      tasks.complete + tasks.rejected;
    if (total > 0) {
      lines.push("");
      lines.push(
        `Tasks: ${tasks.pending} pending, ${tasks.current} current, ${tasks.review} review, ${tasks.complete} complete, ${tasks.rejected} rejected`,
      );
    }
  }

  return lines.join("\n");
}

File: src/cmd/status.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/AdriaanVE/jack-in/src/config"
	"github.com/AdriaanVE/jack-in/src/daemon"
	"github.com/AdriaanVE/jack-in/src/domain"
	"github.com/AdriaanVE/jack-in/src/taskqueue"
	"github.com/AdriaanVE/jack-in/src/tmux"
	"github.com/AdriaanVE/jack-in/src/worktree"
	"github.com/spf13/cobra"
)

var statusJSON bool

func init() {
	statusCmd.Flags().BoolVar(&statusJSON, "json", false, "Output as JSON")
	rootCmd.AddCommand(statusCmd)
}

var statusCmd = &cobra.Command{
	Use:   "status",
	Short: "Show worker, daemon, and task status",
	RunE:  runStatus,
}

// shells that indicate a worker process has exited back to shell.
var shells = map[string]bool{"bash": true, "zsh": true, "fish": true, "sh": true}

type workerStatus struct {
	Name   string `json:"name"`
	Agent  string `json:"agent"`
	State  string `json:"state"`
	Task   string `json:"task,omitempty"`
	Branch string `json:"branch,omitempty"`
}

type statusOutput struct {
	Session        string            `json:"session"`
	Uptime         string            `json:"uptime,omitempty"`
	Approval       string            `json:"approval"`
	ApprovalSource string            `json:"approval_source"`
	Daemon         bool              `json:"daemon"`
	Workers        []workerStatus    `json:"workers"`
	Tasks          domain.TaskCounts `json:"tasks"`
}

func runStatus(cmd *cobra.Command, args []string) error {
	base := cwd()
	cfg, err := loadConfig(base)
	if err != nil {
		return err
	}
	session := config.SessionName(cfg.Project)

	if !tmux.HasSession(session) {
		if statusJSON {
			out, _ := json.MarshalIndent(statusOutput{Session: session}, "", "  ")
			fmt.Println(string(out))
			return nil
		}
		fmt.Printf("Session '%s' is not running. Run 'jackin up' first.\n", session)
		return nil
	}

	panes, err := tmux.ListPanes(session)
	if err != nil {
		return fmt.Errorf("listing panes: %w", err)
	}
	paneByWindow := make(map[string]tmux.PaneInfo)
	for _, p := range panes {
		paneByWindow[p.WindowName] = p
	}

	// Detect daemon: pane 0 of dashboard-orchestrator running a non-shell command.
	// Target pane 0 explicitly because the window has two panes (daemon + orchestrator)
	// and the paneByWindow map only retains the last pane per window name.
	daemonRunning := false
	for _, p := range panes {
		if p.WindowName == "dashboard-orchestrator" && p.PaneIndex == 0 {
			daemonRunning = !p.PaneDead && p.CurrentCommand != "" && !shells[p.CurrentCommand]
			break
		}
	}

	var uptimeStr string
	if created, err := tmux.SessionCreated(session); err == nil {
		uptimeStr = time.Since(time.Unix(created, 0)).Truncate(time.Second).String()
	}

	approval, approvalSource := resolveApproval(base, cfg)

	counts, countsErr := taskqueue.Counts(base)
	if countsErr != nil {
		fmt.Fprintf(os.Stderr, "warning: reading task counts: %v\n", countsErr)
	}

	workers := make([]workerStatus, 0, len(cfg.Workers))
	for _, w := range cfg.Workers {
		ws := workerStatus{
			Name:   w.Name,
			Agent:  string(w.Agent),
			Branch: worktree.BranchName(cfg.Project, w.Name),
		}

		p, paneExists := paneByWindow[w.Name]
		switch {
		case !paneExists:
			ws.State = "gone"
		case p.PaneDead || shells[p.CurrentCommand]:
			ws.State = "stopped"
		default:
			taskData, err := os.ReadFile(daemon.CurrentTaskFilePath(base, w.Name))
			if err == nil && len(taskData) > 0 {
				ws.State = "working"
				ws.Task = string(taskData)
			} else {
				ws.State = "waiting"
			}
		}

		workers = append(workers, ws)
	}

	out := statusOutput{
		Session:        session,
		Uptime:         uptimeStr,
		Approval:       string(approval),
		ApprovalSource: approvalSource,
		Daemon:         daemonRunning,
		Workers:        workers,
		Tasks:          counts,
	}

	if statusJSON {
		b, _ := json.MarshalIndent(out, "", "  ")
		fmt.Println(string(b))
		return nil
	}

	// Human-readable output
	fmt.Printf("Session:  %s", session)
	if uptimeStr != "" {
		fmt.Printf("  (up %s)", uptimeStr)
	}
	fmt.Println()

	fmt.Printf("Daemon:   %s\n", boolLabel(daemonRunning, "running", "stopped"))
	approvalLabel := string(approval)
	if approval == domain.ApprovalAuto {
		model := os.Getenv("ANTHROPIC_DEFAULT_SONNET_MODEL")
		if model == "" {
			model = "claude-sonnet-4-5"
		}
		approvalLabel += " (" + model + ")"
	}
	fmt.Printf("Approval: %s (%s)\n", approvalLabel, approvalSource)

	fmt.Println()
	fmt.Println("Workers:")
	for _, w := range workers {
		line := fmt.Sprintf("  %-12s %-10s %s", w.Name, w.Agent, w.State)
		if w.Task != "" {
			line += "  task=" + w.Task
		}
		fmt.Println(line)
	}

	fmt.Println()
	fmt.Printf("Tasks: %d pending, %d current, %d review, %d complete, %d rejected\n",
		counts.Pending, counts.Current, counts.Review, counts.Complete, counts.Rejected)

	return nil
}

func boolLabel(b bool, trueLabel, falseLabel string) string {
	if b {
		return trueLabel
	}
	return falseLabel
}
